Spartan docs (#16)
anonklub/docs: docs/prove/spartan.md

> > Refer to the [Spartan circuit in `anonklub/anonklub` repository](https://github.com/anonklub/anonklub/tree/main/pkgs)

Reviewer: ```suggestion
> Refer to the [Spartan circuit in `anonklub/anonklub` repository](https://github.com/anonklub/anonklub/tree/main/pkgs/spartan-ecdsa-wasm/src)
```
?

export const useSpartanEcdsaWorker = () => {
  const isWorkerReady = useWorker(SpartanEcdsaWorker)

  const proveMembership: ProveMembershipFn = async ({
    merkleProofBytesSerialized,
    message,
    sig,
  }): Promise<Uint8Array> => {
    process.env.NODE_ENV === 'development' && console.time('==> Prove')

    const proof = await SpartanEcdsaWorker.proveMembership({
      merkleProofBytesSerialized,
      message,
      sig,
    })

    process.env.NODE_ENV === 'development' && console.timeEnd('==> Prove')

    return proof
  }

  const verifyMembership: VerifyMembershipFn = async (
    anonklubProof: Uint8Array,
  ): Promise<boolean> => {
    process.env.NODE_ENV === 'development' && console.time('==> Verify')

    const isVerified = await SpartanEcdsaWorker.verifyMembership(anonklubProof)

    process.env.NODE_ENV === 'development' && console.timeEnd('==> Verify')

    return isVerified
  }

  return {
    isWorkerReady,
    proveMembership,
    verifyMembership,
  }
}
```

## Step By Step

### Prepare

> [@anonklub/merkle-tree-worker](#) and [@anonklub/spartan-ecdsa-worker](#) are designed to operate on the client side. In the example above, ensure that you prepare each worker using `await worker.prepare()`. Please check `Wasm` doc for more details.

Reviewer: ```
 Please check [`wasm` doc](LINK??) for more details.
```

## Merkle Proof

> Generating a Merkle proof to verify the inclusion of an Ethereum address within a Merkle tree is crucial for the circuit's functionality. We use a binary Merkle tree structure `@anonklub/merkle-tree-worker` library in case of Spartan circuits. Follow these steps to generate a Merkle proof:

Reviewer: ```suggestion
Generating a Merkle proof to verify the inclusion of an Ethereum address within a Merkle tree is crucial for the circuit's functionality. We use a the `@anonklub/merkle-tree-worker` library to run the WebAssembly compilation of our merkle tree rust implementation ([`@anonklub/merkle-tree-wasm`](https://www.npmjs.com/package/@anonklub/merkle-tree-wasm)) in a web worker. Follow these steps to generate a Merkle proof:
```

not a binary tree here, let's include links too

Author: > not a binary tree here
https://github.com/anonklub/anonklub.github.io/pull/16#discussion_r1756216226

> let's include links too
But I can include links here.

> Generating a Merkle proof to verify the inclusion of an Ethereum address within a Merkle tree is crucial for the circuit's functionality. We use a binary Merkle tree structure `@anonklub/merkle-tree-worker` library in case of Spartan circuits. Follow these steps to generate a Merkle proof:

Reviewer: In this case
```suggestion
Generating a Merkle proof to verify the inclusion of an Ethereum address within a Merkle tree is crucial for the circuit's functionality. We use a the `@anonklub/merkle-tree-worker` library to run the WebAssembly compilation of our binary merkle tree rust implementation ([`@anonklub/merkle-tree-wasm`](https://www.npmjs.com/package/@anonklub/merkle-tree-wasm)) in a web worker. Follow these steps to generate a Merkle proof:
```

1. Call the `prepare()` function as outlined earlier.
2. Sign a `message`and obtain the `signature` in hexadecimal format.
3. The generated Spartan membership proof `membershipProofSerialized` is in a serialized form and it will be ready for immediate use in the `verifyMembership()` function.

```js
export interface ProveInputs {
  sig: Hex
  message: string
  merkleProofBytesSerialized: Uint8Array
}
```

### Example of use

```js
import { useAsync } from 'react-use'
import type { Hex } from 'viem'
import { useSpartanEcdsaWorker } from './useSpartanEcdsaWorker'
import { useStore } from './useStore'

export const useProofResult = () => {
  const { proofRequest } = useStore()
  const { isWorkerReady, proveMembership } = useSpartanEcdsaWorker()

  return useAsync(async () => {
    if (proofRequest === null || !isWorkerReady) return

    return await proveMembership({
      merkleProofBytesSerialized: proofRequest.merkleProof,
      message: proofRequest.message,
      sig: proofRequest.rawSignature as Hex,
    })
  }, [isWorkerReady, proofRequest])
}
```

> ## Verify of Membership

Author: I even changed `Proof of Membership` to `Membership proof` to keep both the same style.

> After successfully generating the Spartan proof, you can proceed with verifying that proof in Spartan.

Reviewer: "verifying that proof _in_ Spartan" sounds weird to me.
what about simply
"verifying that proof ~~in Spartan~~"
(we are on the spartan doc page, it is clear already with deal with spartan here)

> Refer to the [Spartan circuit in `anonklub/anonklub` repository](https://github.com/anonklub/anonklub/tree/main/pkgs)
> This work here is mainly based on [spartan-ecdsa](https://github.com/personaelabs/spartan-ecdsa) from Personae labs.

> Requirements
>

> > - [ ] Install [@anonklub/merkle-tree-worker](https://www.npmjs.com/package/@anonklub/merkle-tree-worker), a web worker that includes the WebAssembly compilation a binary merkle .

Reviewer: what did you want to say?
```suggestion
> - [ ] Install [@anonklub/merkle-tree-worker](https://www.npmjs.com/package/@anonklub/merkle-tree-worker), a web worker that includes the WebAssembly compilation of a [binary merkle tree rust implementation](https://github.com/anonklub/anonklub/tree/main/pkgs/merkle-tree-wasm).
```

> > - [ ] Install [@anonklub/spartan-ecdsa-worker](https://www.npmjs.com/package/@anonklub/spartan-ecdsa-worker), a web worker that contains the WebAssembly compilation of the Spartan circuit.

Reviewer: ```suggestion
> - [ ] Install [@anonklub/spartan-ecdsa-worker](https://www.npmjs.com/package/@anonklub/spartan-ecdsa-worker), a web worker that contains the WebAssembly compilation of the Spartan circuit, implementated with the [`sapir`](https://github.com/personaelabs/sapir) rust crate.
```

## TLDR

### 

> ```js

Reviewer: can you please make clearer here that these are React hooks.
Then add a code snippet that shows how to use these hooks in a dummy component (or link to the real React components used in our UI app)

// React.js / Next.js example
import { type ProveMembershipFn, SpartanEcdsaWorker, type VerifyMembershipFn } from '@anonklub/spartan-ecdsa-worker'
import { type GenerateMerkleProofFn, MerkleTreeWorker } from '@anonklub/merkle-tree-worker'
import { useEffect, useState } from 'react'

export const useWorker = (
  worker: typeof SpartanEcdsaWorker | typeof MerkleTreeWorker
) => {
  const [isWorkerReady, setIsWorkerReady] = useState(false)

  useEffect(() => {
    void (async () => {
      await worker.prepare()
      setIsWorkerReady(true)
    })()

>     // eslint-disable-next-line react-hooks/exhaustive-deps

Reviewer: let's remove the eslint statements in all this page.
Let's not assume the reader will use eslint

      return proof
    } catch (error) {
      throw error
    }
  }

  return {
    generateMerkleProof,
    isWorkerReady,
  }
}

export const useSpartanEcdsaWorker = () => {
  const isWorkerReady = useWorker(SpartanEcdsaWorker)

  const proveMembership: ProveMembershipFn = async ({
    merkleProofBytesSerialized,
    message,
    sig,
  }): Promise<Uint8Array> => {
    process.env.NODE_ENV === 'development' && console.time('==> Prove')

    const proof = await SpartanEcdsaWorker.proveMembership({
      merkleProofBytesSerialized,
      message,
      sig,
    })

    process.env.NODE_ENV === 'development' && console.timeEnd('==> Prove')

    return proof
  }

  const verifyMembership: VerifyMembershipFn = async (
    anonklubProof: Uint8Array,
  ): Promise<boolean> => {
    process.env.NODE_ENV === 'development' && console.time('==> Verify')

    const isVerified = await SpartanEcdsaWorker.verifyMembership(anonklubProof)

    process.env.NODE_ENV === 'development' && console.timeEnd('==> Verify')

    return isVerified
  }

  return {
    isWorkerReady,
    proveMembership,
    verifyMembership,
  }
}
```

## Step By Step

> ### Prepare

Reviewer: same i think this can go into
https://github.com/anonklub/docs/blob/main/docs/prove/wasm.md (see the TODO)

the interface is the same between all the wasm package right ? (merkle, spartan, halo2)

> [@anonklub/merkle-tree-worker](#) and [@anonklub/spartan-ecdsa-worker](#) are designed to operate on the client side. In the example above, ensure that you run prepare from each worker using `await worker.prepare()`. This function initializes the WebAssembly (WASM) code from the underlying rust code:

Reviewer: ```suggestion
[@anonklub/merkle-tree-worker](#) and [@anonklub/spartan-ecdsa-worker](#) are designed to operate on the client side. In the example above, ensure that you prepare each worker using `await worker.prepare()`. This function initializes the WebAssembly (WASM) code from the underlying rust code:
```

#### `MerkleTreeWorker.prepare()`

```js
async prepare() {
    merkleTreeWasm = await import('@anonklub/merkle-tree-wasm')
}
```

#### `SpartanEcdsaWorker.prepare()`

```js
async prepare() {
    spartanEcdsaWasm = await import('@anonklub/spartan-ecdsa-wasm')
    spartanEcdsaWasm.init_panic_hook()

    if (!initialized) {
        spartanEcdsaWasm.prepare()
        initialized = true
    }
}
```

## Merkle Proof

> Generating a Merkle proof to verify the inclusion of an Ethereum address within a Merkle tree is crucial for the circuit's functionality. We utilize a binary Merkle tree structure, and the `@anonklub/merkle-tree-worker` library provides a gadget that serves as a constraint for verifying the Merkle proof within the circuit. Follow these steps to generate a Merkle proof:

Reviewer: ```suggestion
Generating a Merkle proof to verify the inclusion of an Ethereum address within a Merkle tree is crucial for the circuit's functionality. We use a binary Merkle tree structure, and the `@anonklub/merkle-tree-worker` library provides a gadget that serves as a constraint for verifying the Merkle proof within the circuit. Follow these steps to generate a Merkle proof:
```
and can you link to some halo2 docs defining what is a "gadget" or use another term?

Author: That will be added in case of Halo2 doc.

1. Call the `prepare()` function as previously described.

> 2. Prepare the list of Ethereum addresses. The Anonklub project can assist in scanning the blockchain to create this list.

Reviewer: add a link to the docs page about the query API

3. Define the parameters needed to generate the Merkle proof, including the number of `leaves` in the tree, the tree's `depth` (e.g., default: 15), and the specific `leaf` (address) for which you want to prove membership.
4. The generated proof will be serialized, making it immediately usable as a parameter for the proof function in the circuit.

```js
export type GenerateMerkleProofFn = (
  leaves: string[],
  leaf: string,
  depth: number,
) => Promise<Uint8Array>
```

## Proof of Membership

> Once you have generated the Merkle proof for an Ethereum address (leaf), you can proceed to create a Spartan proof for the membership of that address in the Merkle tree. Follow these steps to generate a Spartan membership proof:

Reviewer: ```suggestion
Once you have generated the Merkle proof for an Ethereum address (leaf), you can proceed with creating a Spartan proof for the membership of that address in the Merkle tree. Follow these steps to generate a Spartan membership proof:
```
